### PR TITLE
Add functions to check elemental host and elemental machine resources status

### LINKED ISSUE #14: Improve check in the CI
Add more checks in the CI like `elementalmachine` and `elementalhost` status before going forward in the test.
https://github.com/rancher-sandbox/elemental-e2e/blob/main/tests/e2e/bootstrap_capi_test.go#L148-L162

### PR DESCRIPTION
**Pull request by juadk**
Fix #14 

Before ending the test, we must ensure every elemental resource (host/machine) is ready.

## Verification run
https://github.com/rancher-sandbox/elemental-e2e/actions/runs/9549984180/job/26320936384 ✅

### CODE REVIEW

File: tests/e2e/bootstrap_capi_test.go [hunk 1]
```diff
@@ -145,21 +146,26 @@ var _ = Describe("E2E - Bootstrapping node", Label("bootstrap"), func() {
...
-			rke2-control-plane-jz896   rke2      m-b9de3abc-378d-46d4-989c-5594215f6c7f   elemental://default/m-b9de3abc-378d-46d4-989c-5594215f6c7f   Running   22m   v1.30.1+rke2r1
-			rke2-md-0-tmq6b-4xpxz      rke2      m-2b2d48a9-4401-4b59-9553-3aaa4856e5e0   elemental://default/m-2b2d48a9-4401-4b59-9553-3aaa4856e5e0   Running   22m   v1.30.1+rke2r1
-		*/
+		By("Checking elemental hosts status", func() {
+			for index := vmIndex; index <= numberOfVMs; index++ {
+				// Set node hostname
+				hostName := elemental.SetHostname(vmNameRoot, index)
+				Expect(hostName).To(Not(BeEmpty()))
+				GinkgoWriter.Printf("Check elementalhost %s\n", hostName)
+				WaitElementalHost("default", hostName)
```
Comment (ldevulder) [inline tests/e2e/bootstrap_capi_test.go:155]: Please avoid hardcoded Namespace ;-)

File: tests/e2e/bootstrap_capi_test.go [hunk 2]
```diff
@@ -145,21 +146,26 @@ var _ = Describe("E2E - Bootstrapping node", Label("bootstrap"), func() {
...
+		})
+
+		By("Checking elemental machines status", func() {
+			elementalMachineList, err := kubectl.RunWithoutErr("get", "elementalmachine",
+				"--namespace", clusterNS, "-o", "jsonpath={.items[*].metadata.name}")
+			Expect(err).To(Not(HaveOccurred()))
+
+			for _, machine := range strings.Fields(elementalMachineList) {
+				GinkgoWriter.Printf("Check elementalmachine %s\n", machine)
+				WaitElementalMachine("default", machine)
```
Comment (ldevulder) [inline tests/e2e/bootstrap_capi_test.go:166]: Please avoid hardcoded Namespace ;-)

**Review by ldevulder — COMMENTED** [review]
You can try to merge the 2 new functions in one as they are very closed, In that case I think adding some thing like a `stateToCheck` variable to define the `states := []state{ [...]` variable could be good. Not a good idea to check the whole `conditionStatus` in all cases.

File: tests/e2e/bootstrap_capi_test.go (hunk 2 above)
Comment (juadk) [reply tests/e2e/bootstrap_capi_test.go]: I know, I have an issue for that, I will fix it in a follow-up PR :)

File: tests/e2e/bootstrap_capi_test.go (hunk 1 above)
Comment (juadk) [reply tests/e2e/bootstrap_capi_test.go]: I know, I have an issue for that, I will fix it in a follow-up PR :)

File: tests/e2e/suite_test.go [hunk 3]
```diff
@@ -119,6 +119,94 @@ func WaitCAPICluster(ns, cn string) {
...
+  - @returns Nothing, the function will fail through Ginkgo in case of issue
+*/
+func WaitElementalResources(ns, er, rs string) {
+	type state struct {
+		conditionStatus string
+		conditionType   string
+	}
+
+	// List of conditions to check
+	elementalhostStates := []state{
```
Comment (ldevulder) [inline tests/e2e/suite_test.go:136]: should be `elementalHostStates`.

File: tests/e2e/suite_test.go [hunk 4]
```diff
@@ -119,6 +119,94 @@ func WaitCAPICluster(ns, cn string) {
...
+			conditionType:   "BootstrapReady",
+		},
+		{
+			conditionStatus: "True",
+			conditionType:   "Ready",
+		},
+	}
+
+	// List of conditions to check
+	elementalmachineStates := []state{
```
Comment (ldevulder) [inline tests/e2e/suite_test.go:156]: Should be `elementalMachineStates`.

File: tests/e2e/suite_test.go [hunk 5]
```diff
@@ -119,6 +119,94 @@ func WaitCAPICluster(ns, cn string) {
...
+			conditionType:   "ProviderIDReady",
+		},
+		{
+			conditionStatus: "True",
+			conditionType:   "Ready",
+		},
+	}
+
+	// Check that all needed conditions are in the good state
+	if er == "elementalhost" {
```
Comment (ldevulder) [inline tests/e2e/suite_test.go:176]: For next change, better to use `switch`, like this (code not tested!):
```go
var statesCheck []state

switch er {
case "elementalhost":
	statesCheck = elementalHostStates
case "elementalmachine":
	statesCheck = elementalMachineStates
}

// Check that all needed conditions are in the good state
for _, s := range statesCheck {
	CheckCondition(ns, er, rs, s.conditionType, s.conditionStatus)
}
```

Another solution could be also to add the definition of the `elemental*States` variables inside the `switch` condition. You can test what is more readable.

**Pull request merged**